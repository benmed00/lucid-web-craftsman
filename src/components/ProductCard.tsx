import { Link } from "react-router-dom";
import { ShoppingCart, Eye, AlertTriangle, Share } from "lucide-react";
import { Product } from "@/shared/interfaces/Iproduct.interface";
import OptimizedImage from "@/components/performance/OptimizedImage";
import { Card, CardContent } from "@/components/ui/card";
import { Badge } from "@/components/ui/badge";
import { Button } from "@/components/ui/button";
import { WishlistButton } from "@/components/ui/WishlistButton";
import { NativeShare } from "@/components/ui/NativeShare";
import { TooltipWrapper } from "@/components/ui/TooltipWrapper";
import { useStock } from "@/hooks/useStock";
import { StockInfo } from "@/services/stockService";
import { useEffect, useState } from "react";
import { useCurrency } from "@/context/CurrencyContext";

interface ProductCardProps {
  product: Product;
  onAddToCart: (product: Product) => void;
  onQuickView?: (product: Product) => void;
}

const ProductCard = ({ product, onAddToCart, onQuickView }: ProductCardProps) => {
  const { stockInfo } = useStock({ productId: product.id });
  const [showShareDialog, setShowShareDialog] = useState(false);
  const { formatPrice } = useCurrency();

  // Type guard to ensure stockInfo is StockInfo for single product
  const singleStockInfo = stockInfo as StockInfo | null;
  
  // Generate unique IDs for accessibility
  const cardId = `product-card-${product.id}`;
  const quickViewBtnId = `quick-view-btn-${product.id}`;
  const addToCartBtnId = `add-to-cart-btn-${product.id}`;
  const wishlistBtnId = `wishlist-btn-${product.id}`;
  const shareBtnId = `share-btn-${product.id}`;

  return (
    <Card 
      id={cardId}
      className="bg-white border border-stone-100 overflow-hidden group hover:shadow-2xl transition-all duration-500 relative touch-manipulation rounded-2xl hover:scale-[1.02] hover:-translate-y-1 shadow-lg hover:border-olive-200"
    >
      <Link to={`/products/${product.id}`} className="block touch-manipulation">
        <div className="relative group/image">
          <div className="aspect-[4/3] w-full overflow-hidden rounded-t-xl">
            <OptimizedImage
              src={product.images[0]}
              alt={product.name}
              width={400}
              height={300}
              aspectRatio="4/3"
              className="object-cover w-full h-full group-hover:scale-110 transition-transform duration-700"
              priority={false}
              webp={true}
              quality={85}
            />
          </div>
          
          {/* Quick View Button - repositioned to top-right of image */}
          {onQuickView && (
            <div className="absolute top-3 right-3 z-20">
              <TooltipWrapper 
                content={`Aperçu rapide de ${product.name}`}
                side="left"
              >
                <Button
                  id={quickViewBtnId}
                  name={`quick-view-${product.name.toLowerCase().replace(/\s+/g, '-')}`}
                  size="sm"
                  variant="secondary"
                  onClick={(e) => {
                    e.preventDefault();
                    e.stopPropagation();
                    onQuickView(product);
                  }}
                  className="bg-white/95 backdrop-blur-sm hover:bg-white p-2.5 rounded-full shadow-lg hover:shadow-xl transition-all duration-200 hover:scale-110"
                  aria-label={`Aperçu rapide de ${product.name}`}
                >
                  <Eye className="h-4 w-4" />
                </Button>
              </TooltipWrapper>
            </div>
          )}

          {/* Share Button - Mobile Only */}
          <div className="absolute top-16 right-3 z-10 md:hidden">
            <TooltipWrapper 
              content={`Partager ${product.name}`}
              side="left"
            >
              <Button
                id={shareBtnId}
                name={`share-${product.name.toLowerCase().replace(/\s+/g, '-')}`}
                size="sm"
                variant="secondary"
                onClick={(e) => {
                  e.preventDefault();
                  e.stopPropagation();
                  setShowShareDialog(true);
                }}
                className="bg-white/90 backdrop-blur-sm hover:bg-white shadow-lg p-2 rounded-full touch-manipulation"
                aria-label={`Partager ${product.name}`}
              >
                <Share className="h-4 w-4" />
              </Button>
            </TooltipWrapper>
          </div>

          {(product.new || product.is_new) && (
            <Badge className="absolute bottom-2 right-2 md:bottom-3 md:right-3 bg-olive-700 text-white border-none shadow-lg text-xs px-2 py-1">
              Nouveau
            </Badge>
          )}

          {/* Stock alerts */}
          {singleStockInfo && (
            <>
              {singleStockInfo.isOutOfStock && (
                <Badge className="absolute top-12 right-2 md:top-14 md:right-3 bg-red-600 text-white border-none shadow-lg text-xs px-2 py-1">
                  {singleStockInfo.message}
                </Badge>
              )}
              {singleStockInfo.isLow && !singleStockInfo.isOutOfStock && (
                <Badge className="absolute top-2 left-2 md:top-3 md:left-3 bg-amber-500 text-white border-none shadow-lg flex items-center gap-1 text-xs px-2 py-1 rounded-full">
                  <AlertTriangle className="h-3 w-3" />
                  <span className="hidden sm:inline">{singleStockInfo.message}</span>
                  <span className="sm:hidden">Il ne reste que {singleStockInfo.maxQuantity || 3} pièces</span>
                </Badge>
              )}
            </>
          )}
        </div>
      </Link>
      
      <CardContent className="p-5 md:p-6 relative">
        {/* Wishlist button - repositioned to top-right of white section */}
        <div 
          className="absolute -top-2 right-4 z-20"
          id={wishlistBtnId}
          data-name={`wishlist-${product.name.toLowerCase().replace(/\s+/g, '-')}`}
        >
          <WishlistButton 
            productId={product.id}
            size="sm"
            variant="ghost"
            className="bg-white border-2 border-stone-200 hover:bg-stone-50 p-2.5 rounded-full shadow-lg hover:shadow-xl transition-all duration-200 hover:scale-110"
          />
        </div>

        <p className="text-sm text-olive-700 font-medium mb-2 uppercase tracking-wide">
          {product.category}
        </p>
        <Link to={`/products/${product.id}`} className="touch-manipulation">
          <h3 className="font-serif text-lg md:text-xl font-medium text-stone-800 mb-4 line-clamp-2 leading-tight hover:text-olive-700 transition-colors duration-200 pr-12">
            {product.name}
          </h3>
        </Link>
        <div className="flex flex-col gap-4">
          <div className="relative">
            <p className="text-stone-800 font-bold text-xl md:text-2xl whitespace-nowrap">
              {formatPrice(product.price)}
            </p>
          </div>
          <TooltipWrapper 
            content={singleStockInfo?.isOutOfStock 
              ? `${product.name} est actuellement indisponible` 
              : `Ajouter ${product.name} à votre panier (${formatPrice(product.price)})`
            }
            side="top"
          >
            <Button
              id={addToCartBtnId}
              name={`add-to-cart-${product.name.toLowerCase().replace(/\s+/g, '-')}`}
              size="lg"
              onClick={(e) => {
                e.preventDefault();
                e.stopPropagation();
                onAddToCart(product);
              }}
              disabled={singleStockInfo?.isOutOfStock}
              className="w-full bg-olive-700 hover:bg-olive-800 text-white hover:text-white active:scale-[0.98] transition-all duration-200 disabled:opacity-50 disabled:cursor-not-allowed px-6 py-4 touch-manipulation min-h-[52px] font-semibold text-base rounded-xl shadow-lg hover:shadow-xl disabled:hover:bg-olive-700 disabled:hover:shadow-lg group relative overflow-hidden border-0"
              aria-label={singleStockInfo?.isOutOfStock 
                ? `${product.name} est indisponible` 
                : `Ajouter ${product.name} au panier pour ${formatPrice(product.price)}`
              }
            >
              <div className="relative flex items-center justify-center gap-2 z-10">
                <ShoppingCart className="h-5 w-5" />
                <span className="font-medium">
                  {singleStockInfo?.isOutOfStock ? 'Indisponible' : 'Ajouter au panier'}
                </span>
              </div>
              
              {/* Hover gradient overlay */}
              <div className="absolute inset-0 bg-gradient-to-r from-olive-600 to-olive-800 opacity-0 group-hover:opacity-100 transition-opacity duration-300" />
            </Button>
          </TooltipWrapper>
        </div>
      </CardContent>

      {/* Native Share Dialog */}
      <NativeShare
        title={product.name}
        text={`Découvrez ce magnifique ${product.name} - ${product.description}`}
        url={`${window.location.origin}/products/${product.id}`}
        isOpen={showShareDialog}
        onClose={() => setShowShareDialog(false)}
      />
    </Card>
  );
};

export default ProductCard;