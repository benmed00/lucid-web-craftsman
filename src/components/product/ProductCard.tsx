import React from 'react';
import { Link } from 'react-router-dom';
import { Card, CardContent } from '@/components/ui/card';
import { Badge } from '@/components/ui/badge';
import { Button } from '@/components/ui/button';
import { ShoppingBag } from 'lucide-react';
import { Product } from '@/shared/interfaces/Iproduct.interface'; // Adjusted path

interface ProductCardProps {
  product: Product;
  onAddToCart: (product: Product, event: React.MouseEvent) => void;
}

const ProductCard: React.FC<ProductCardProps> = ({ product, onAddToCart }) => {
  return (
    <Link
      to={`/products/${product.id}`}
      key={product.id} // Key should ideally be on the list item in ProductGrid
      className="group relative"
    >
      <Card className="border-none shadow-sm overflow-hidden hover-scale">
        <div className="aspect-ratio aspect-w-1 aspect-h-1 relative overflow-hidden">
          <img
            src={product.images[0]}
            alt={product.name}
            className="object-cover w-full h-full transition-transform duration-500 group-hover:scale-105"
          />
          {product.new && (
            <div className="absolute top-2 right-2">
              <Badge className="bg-olive-500 text-white border-none">
                Nouveau
              </Badge>
            </div>
          )}
        </div>
        <CardContent className="p-4">
          <div className="text-xs text-stone-500 mb-1">
            {product.category}
          </div>
          <h3 className="font-medium text-stone-800 mb-1">
            {product.name}
          </h3>
          <div className="flex items-center justify-between mt-2">
            <p className="text-olive-700 font-medium">
              {product.price} â‚¬
            </p>
            <Button 
              size="sm" 
              className="bg-olive-700 hover:bg-olive-800"
              onClick={(e) => onAddToCart(product, e)}
            >
              <ShoppingBag className="h-4 w-4 mr-1" />
              Ajouter
            </Button>
          </div>
        </CardContent>
      </Card>
    </Link>
  );
};

export default ProductCard;
