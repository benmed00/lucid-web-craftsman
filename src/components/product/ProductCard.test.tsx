// src/components/product/ProductCard.test.tsx
import { describe, it, expect, vi } from 'vitest';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import { BrowserRouter } from 'react-router-dom'; // Needed because ProductCard uses <Link>
import ProductCard from './ProductCard';
import { Product } from '@/shared/interfaces/Iproduct.interface'; // Adjust path as necessary

const mockProduct: Product = {
  id: 1,
  name: 'Sac Artisanal',
  category: 'Sacs',
  price: 75,
  images: ['/test-image.jpg'],
  description: 'Un beau sac fait main.',
  new: true,
  artisan: 'Artisan Test',
  artisanStory: 'Story here',
  details: 'Details here',
  care: 'Care here',
  related: []
};

const mockProductNotNew: Product = {
  ...mockProduct,
  id: 2,
  name: 'Chapeau Simple',
  new: false,
};

const mockOnAddToCart = vi.fn();

// Wrapper component to provide Router context
const renderWithRouter = (ui: React.ReactElement) => {
  return render(ui, { wrapper: BrowserRouter });
};

describe('ProductCard', () => {
  beforeEach(() => {
    mockOnAddToCart.mockClear();
  });

  it('renders product information correctly', () => {
    const { container } = renderWithRouter(
      <ProductCard product={mockProduct} onAddToCart={mockOnAddToCart} />
    );

    expect(screen.getByText('Sac Artisanal')).toBeInTheDocument();
    expect(screen.getByText('Sacs')).toBeInTheDocument();
    expect(screen.getByText(/75\s*â‚¬/)).toBeInTheDocument(); // Regex for price with euro symbol
    expect(screen.getByAltText('Sac Artisanal')).toHaveAttribute('src', '/test-image.jpg');
    expect(screen.getByText('Nouveau')).toBeInTheDocument(); // Test for "new" badge

    // Optional: Snapshot test
    // expect(container).toMatchSnapshot();
  });

  it('does not render "new" badge if product.new is false', () => {
    renderWithRouter(
      <ProductCard product={mockProductNotNew} onAddToCart={mockOnAddToCart} />
    );
    expect(screen.queryByText('Nouveau')).not.toBeInTheDocument();
  });

  it('handles onAddToCart click', async () => {
    const user = userEvent.setup();
    renderWithRouter(
      <ProductCard product={mockProduct} onAddToCart={mockOnAddToCart} />
    );

    const addButton = screen.getByRole('button', { name: /ajouter/i });
    await user.click(addButton);

    expect(mockOnAddToCart).toHaveBeenCalledTimes(1);
    // The first argument to onAddToCart is the product, the second is the event.
    // We can check if it was called with the product.
    expect(mockOnAddToCart.mock.calls[0][0]).toEqual(mockProduct);
  });

  // Example of a snapshot test (uncomment to use)
  it('matches snapshot', () => {
    const { container } = renderWithRouter(
      <ProductCard product={mockProduct} onAddToCart={mockOnAddToCart} />
    );
    expect(container).toMatchSnapshot();
  });
});
