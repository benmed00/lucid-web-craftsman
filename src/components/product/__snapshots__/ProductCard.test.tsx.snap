// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`ProductCard > matches snapshot 1`] = `
<div>
  <a
    class="group relative"
    href="/products/1"
  >
    <div
      class="rounded-lg border bg-card text-card-foreground border-none shadow-sm overflow-hidden hover-scale"
    >
      <div
        class="aspect-ratio aspect-w-1 aspect-h-1 relative overflow-hidden"
      >
        <img
          alt="Sac Artisanal"
          class="object-cover w-full h-full transition-transform duration-500 group-hover:scale-105"
          src="/test-image.jpg"
        />
        <div
          class="absolute top-2 right-2"
        >
          <div
            class="inline-flex items-center rounded-full border px-2.5 py-0.5 text-xs font-semibold transition-colors focus:outline-none focus:ring-2 focus:ring-ring focus:ring-offset-2 border-transparent hover:bg-primary/80 bg-olive-500 text-white border-none"
          >
            Nouveau
          </div>
        </div>
      </div>
      <div
        class="p-4"
      >
        <div
          class="text-xs text-stone-500 mb-1"
        >
          Sacs
        </div>
        <h3
          class="font-medium text-stone-800 mb-1"
        >
          Sac Artisanal
        </h3>
        <div
          class="flex items-center justify-between mt-2"
        >
          <p
            class="text-olive-700 font-medium"
          >
            75
             â‚¬
          </p>
          <button
            class="inline-flex items-center justify-center gap-2 whitespace-nowrap text-sm font-medium ring-offset-background transition-colors focus-visible:outline-none focus-visible:ring-2 focus-visible:ring-ring focus-visible:ring-offset-2 disabled:pointer-events-none disabled:opacity-50 [&_svg]:pointer-events-none [&_svg]:size-4 [&_svg]:shrink-0 text-primary-foreground h-9 rounded-md px-3 bg-olive-700 hover:bg-olive-800"
          >
            <svg
              class="lucide lucide-shopping-bag h-4 w-4 mr-1"
              fill="none"
              height="24"
              stroke="currentColor"
              stroke-linecap="round"
              stroke-linejoin="round"
              stroke-width="2"
              viewBox="0 0 24 24"
              width="24"
              xmlns="http://www.w3.org/2000/svg"
            >
              <path
                d="M6 2 3 6v14a2 2 0 0 0 2 2h14a2 2 0 0 0 2-2V6l-3-4Z"
              />
              <path
                d="M3 6h18"
              />
              <path
                d="M16 10a4 4 0 0 1-8 0"
              />
            </svg>
            Ajouter
          </button>
        </div>
      </div>
    </div>
  </a>
</div>
`;
